st2actions/st2actions/utils/param_utils.py

```diff
@@ -45,295 +43,158 @@ def pf(params, skips):
-            if isinstance(renderable_v, dict) or isinstance(renderable_v, list):
-                renderable_v = json.dumps(renderable_v)
-            # only str can contain templates
-            if (isinstance(renderable_v, str) or isinstance(renderable_v, unicode)) and \
-               _is_template(renderable_v):
-                renderable_params[k] = renderable_v
-            elif isinstance(v, dict) or isinstance(v, list):
-                # For context use the renderable value for dict and list params. The template
-                # rendering by jinja yields a non json.loads compatible value leading to issues
-                # while performing casts.
-                context_params[k] = renderable_v
-            else:
-                # For context use the original value.
-                context_params[k] = v
-
-    do_render_context_split(action_parameters)
-    do_render_context_split(runner_parameters)
-
-    return (renderable_params, context_params)
-
-
-def _check_availability(param, param_dependencies, renderable_params, context):
-    for dependency in param_dependencies:
-        if dependency not in renderable_params and dependency not in context:
-            return False
-    return True
-
-
-def _check_cyclic(dep_chain, dependencies):
-    last_idx = len(dep_chain) - 1
-    last_value = dep_chain[last_idx]
-    for dependency in dependencies.get(last_value, []):
-        if dependency in dep_chain:
-            dep_chain.append(dependency)
-            return False
-        dep_chain.append(dependency)
-        if not _check_cyclic(dep_chain, dependencies):
-            return False
-        dep_chain.pop()
-    return True
-
-
-def _validate_dependencies(renderable_params, context):
-    '''
-    Validates dependencies between the parameters.
-    e.g.
-    {
-        'a': '{{b}}',
-        'b': '{{a}}'
-    }
-    In this example 'a' requires 'b' for template rendering and vice-versa. There is no way for
-    these templates to be rendered and will be flagged with an ActionRunnerException.
-    '''
-    env = Environment(undefined=StrictUndefined)
-    dependencies = {}
-    for k, v in six.iteritems(renderable_params):
-        template_ast = env.parse(v)
-        dependencies[k] = meta.find_undeclared_variables(template_ast)
-
-    for k, v in six.iteritems(dependencies):
-        if not _check_availability(k, v, renderable_params, context):
-            msg = 'Dependecy unsatisfied - %s: %s.' % (k, v)
-            raise actionrunner.ActionRunnerException(msg)
-        dep_chain = []
-        dep_chain.append(k)
-        if not _check_cyclic(dep_chain, dependencies):
-            msg = 'Cyclic dependecy found - %s.' % dep_chain
+    for name in G.nodes():
+        if 'value' not in G.node[name] and 'template' not in G.node[name]:
+            msg = 'Dependecy unsatisfied in %s' % name
             raise actionrunner.ActionRunnerException(msg)
 
+    if not nx.is_directed_acyclic_graph(G):
```

Comment:
This can be a separate method. Would lead some automatic documentation. 

Also, I personally find long method hard to read even though I have been responsible for a fair few long methods myself.